internal/mutators/podtemplatespec.go

```diff
@@ -0,0 +1,118 @@
+	pts := *inputPts.DeepCopy()
+
+	if err := p.setAndValidateResourceRequirements(pts.Spec.InitContainers, limitRangeMemory); err != nil {
+		return pts, err
+	}
+
+	if err := p.setAndValidateResourceRequirements(pts.Spec.Containers, limitRangeMemory); err != nil {
+		return pts, err
+	}
+
+	return pts, nil
+}
+
+func (p PodTemplateSpec) setAndValidateResourceRequirements(containers []corev1.Container, limitRangeMemory limitrange.Config) error {
+	for idx := range containers {
+		container := &containers[idx]
+		p.setMemoryRequest(container, limitRangeMemory)
+
+		if err := p.setMemoryLimit(container, limitRangeMemory); err != nil {
+			return err
+		}
+
+		if err := p.validateMemoryRequirements(*container, limitRangeMemory); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p PodTemplateSpec) validateMemoryRequirements(container corev1.Container, limitRangeMemory limitrange.Config) error {
+	memoryRequest := container.Resources.Requests.Memory()
+	memoryLimit := container.Resources.Limits.Memory()
+
+	if memoryRequest.IsZero() || memoryLimit.IsZero() {
+		return fmt.Errorf("memory request (%s) and limit (%s) must be set", memoryRequest.String(), memoryLimit.String())
+	}
+
+	if memoryLimit.Cmp(*memoryRequest) == -1 {
+		return fmt.Errorf("memory limit (%s) must be greater than request (%s)", memoryLimit.String(), memoryRequest.String())
+	}
+
+	if limitRangeMemory.HasMaxLimitRequestRatio {
+		ratio := quantity.Div(*memoryLimit, *memoryRequest, infScaleMicro)
+		if ratio.Cmp(limitRangeMemory.MaxLimitRequestRatio) == 1 {
+			return fmt.Errorf("memory limit (%s) to request (%s) ratio (%s) exceeds MaxLimitRequestRatio (%s)",
+				memoryLimit.String(), memoryRequest.String(), ratio.String(), limitRangeMemory.MaxLimitRequestRatio.String())
+		}
+	}
+
+	return nil
+}
+
+func (p PodTemplateSpec) setMemoryRequest(container *corev1.Container, limitRangeMemory limitrange.Config) {
+	memoryRequest := container.Resources.Requests.Memory()
+	if !memoryRequest.IsZero() || !limitRangeMemory.HasDefaultRequest {
+		return
+	}
+
+	if container.Resources.Requests == nil {
+		container.Resources.Requests = corev1.ResourceList{}
+	}
+	container.Resources.Requests[corev1.ResourceMemory] = limitRangeMemory.DefaultRequest
+}
+
+func (p PodTemplateSpec) setMemoryLimit(container *corev1.Container, limitRangeMemory limitrange.Config) error {
+	memoryRequest := container.Resources.Requests.Memory()
+	memoryLimit := container.Resources.Limits.Memory()
+
+	if !memoryLimit.IsZero() {
+		return nil
+	}
+
+	if container.Resources.Limits == nil {
+		container.Resources.Limits = corev1.ResourceList{}
+	}
+
+	var calculatedLimit resource.Quantity
+
+	if limitRangeMemory.HasMaxLimitRequestRatio && !memoryRequest.IsZero() {
+		calculatedLimit = quantity.Mul(*memoryRequest, limitRangeMemory.MaxLimitRequestRatio)
+	} else {
+		ratioMemoryLimit, err := quantity.MulFloat64(*memoryRequest, p.defaultMemoryLimitRequestRatio)
+		if err != nil {
+			return err
+		}
```

Comment:
What is the error that could come out of this? If it is an overflow or calculation error we are responsible for it seems like we should be able to recover by changing the scale or rounding. 

I think it would be a bad user experience to have requests rejected because of a calculation or parsing error from this library.

Reply:
Right good point.

I've changed `defaultMemoryLimitRequestRatio` to be type `resource.Quantity` so that a pure Mul(quantity, quantity) can be done, no risk of overflow or parsing error.
The parse from float64 to Quantity `defaultMemoryLimitRequestRatio` is done with the option `WithDefaultMemoryLimitRequestRatio` which will panic on parse error. This makes sense so that our app panics immediately if given an un-parsable value.

And the function `setMemoryLimit` no longer returns error